### PR TITLE
feat: add transaction review service and user interface

### PR DESCRIPTION
**Pull request by simonheys**
Adds transaction review service with Storybook for associated components.

This will render a raw JSON response below the review in the UI in dev (to be removed before merge)

Adds env var `ARGENT_TRANSACTION_REVIEW_API_BASE_URL`, you will also need VPN to use the current service.

Awaiting examples of expected responses from API before completion.

https://user-images.githubusercontent.com/175607/176168050-98530e7c-4c2d-4606-9dd1-8691920341e1.mov

### CODE REVIEW

File: packages/extension/package.json [hunk 1]
```diff
@@ -62,6 +62,7 @@
     "@tippyjs/react": "^4.2.6",
     "bignumber.js": "^9.0.2",
     "buffer": "^6.0.3",
+    "colord": "^2.9.2",
```
Comment (simonheys) [inline packages/extension/package.json:65]: Small dependency for simple color value manipulation, e.g. making a hex color value lighter or darker

File: packages/extension/src/ui/features/actions/ConfirmScreen.tsx [hunk 2]
```diff
@@ -83,6 +88,7 @@ export const ConfirmScreen: FC<ConfirmScreenProps> = ({
   confirmButtonText = "Confirm",
   confirmButtonDisabled,
   confirmButtonBackgroundColor,
```
Comment (simonheys) [inline packages/extension/src/ui/features/actions/ConfirmScreen.tsx:90]: `confirmBackgroundColor` is left as it's widely used already elsewhere - should be refactored to use variants at a later date

File: packages/extension/src/shared/transactionReview.service.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
+import { isArray, isString } from "lodash-es"
+import { Call } from "starknet"
+import urlJoin from "url-join"
+
+import { fetcher } from "./utils/fetcher"
+
+const ARGENT_TRANSACTION_REVIEW_API_BASE_URL = process.env
+  .ARGENT_TRANSACTION_REVIEW_API_BASE_URL as string
```
Comment (janek26) [inline packages/extension/src/shared/transactionReview.service.ts:8]: uff, is that what prettier comes up with? Does it work properly with webpack?

File: packages/extension/src/shared/transactionReview.service.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  transactions: Call | Call[]
+}
+
+export const fetchTransactionReview = ({
+  network,
+  accountAddress,
+  transactions,
+}: IFetchTransactionReview) => {
+  if (!ARGENT_TRANSACTION_REVIEW_STARKNET_URL) {
+    return
```
Comment (janek26) [inline packages/extension/src/shared/transactionReview.service.ts:83]: I would probably throw here and catch it later

As nether an actual error nor this "error" should break the application

File: packages/extension/src/ui/components/Button.tsx [hunk 5]
```diff
@@ -1,14 +1,45 @@
+import { colord } from "colord"
```
Comment (janek26) [inline packages/extension/src/ui/components/Button.tsx:1]: I like

File: packages/extension/src/ui/features/actions/ApproveTransactionScreen.tsx [hunk 6]
```diff
@@ -40,7 +45,17 @@ const LeftPaddedField = styled.div`
   margin-left: 8px;
 `
 
-const DISPLAY_RAW_TRANSACTION = false
+const isDev = process.env.NODE_ENV === "development"
+
+const DISPLAY_RAW_TRANSACTION = true
```
Comment (janek26) [inline packages/extension/src/ui/features/actions/ApproveTransactionScreen.tsx:50]: should we remove this whole dev stuff before merging?
At least we should disable it if you want to keep it

File: packages/extension/src/ui/features/actions/transaction/TransactionsList.tsx [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+import { ApiTransactionReviewResponse } from "../../../../shared/transactionReview.service"
+import { WarningIcon } from "../../../components/Icons/WarningIcon"
+import { TokenDetails } from "../../accountTokens/tokens.state"
+import { TransactionBanner } from "./TransactionBanner"
+import { TransactionItem } from "./TransactionItem"
+
+export interface ITransactionsList {
+  transactions: Call | Call[]
+  transactionReview: ApiTransactionReviewResponse | undefined
+  tokensByNetwork: TokenDetails[] | undefined
```
Comment (janek26) [inline packages/extension/src/ui/features/actions/transaction/TransactionsList.tsx:14]:
```suggestion
  tokensByNetwork?: TokenDetails[]
```

File: packages/extension/src/ui/features/actions/transaction/TransactionsList.tsx [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+}) => {
+  const transactionsArray: Call[] = useMemo(
+    () => (isArray(transactions) ? transactions : [transactions]),
+    [transactions],
+  )
+  const transactionReviewWarn = transactionReview?.assessment === "warn"
+  const transactionReviewWarnAssessment =
+    transactionReview?.reason === "recipientIsTokenAddress"
+      ? "You are sending tokens to their own address. This is likely to burn them."
+      : "This transaction has been flagged as dangerous. We recommend you reject this transaction unless you are sure."
```
[the remark this answers is not in the document]
Comment (simonheys) [reply packages/extension/src/ui/features/actions/transaction/TransactionsList.tsx:32]: Yes it should, I agree

File: packages/extension/src/shared/transactionReview.service.ts (hunk 3 above)
Comment (simonheys) [reply packages/extension/src/shared/transactionReview.service.ts:8]: Yes that is 'prettier'. Ends up fine in the bundle - I guess although it looks weird to a mere human, the AST of the code will be the same as if it was all on a single line.

image

Comment (janek26) [reply packages/extension/src/shared/transactionReview.service.ts:8]: cool! thanks for confirming, I had these type of web pack bugs before and once you have them you dont know what's wrong.

File: packages/extension/src/ui/features/accounts/UpgradeScreen.tsx [hunk 9]
```diff
@@ -1,11 +1,12 @@
-import { FC, useEffect } from "react"
+import { FC, useCallback, useEffect } from "react"
 import { useNavigate } from "react-router-dom"
 import styled from "styled-components"
 
 import { useAppState } from "../../app.state"
```
Comment (janek26) [inline packages/extension/src/ui/features/accounts/UpgradeScreen.tsx:5]: never used

File: packages/extension/src/ui/features/accounts/UpgradeScreen.tsx [hunk 10]
```diff
@@ -1,11 +1,12 @@
...
 import { P } from "../../components/Typography"
 import { routes } from "../../routes"
 import { upgradeAccount } from "../../services/backgroundAccounts"
```
Comment (janek26) [inline packages/extension/src/ui/features/accounts/UpgradeScreen.tsx:8]: never used

**Review by janek26 — APPROVED** [review]

**Pull request merged**